File: String_Array/ThreeSum/three_sum.js
const threeSum = (arr) => {
  let output = [];
  arr.sort((a, b) => a - b);

  for (let i = 0; i < arr.length - 2; i++) {
    let left = i + 1;
    let right = arr.length - 1;

    if (i > 0 && arr[i] == arr[i - 1]) continue;

    while (left < right) {
      let total = arr[left] + arr[right] + arr[i];
      if (total == 0) {
        output.push([arr[i], arr[left], arr[right]]);

        while (left < right && arr[right - 1] == arr[right]) {
          right--;
        }

        while (left < right && arr[left + 1] == arr[left]) {
          left++;
        }

        right--;
        left++;
      } else if (total > 0) {
        right--;
      } else {
        left++;
      }
    }
  }

  return output;
};


const run = (() => {
  const arr = [-1, 0, 1, 2, -1, -4]
  console.log(threeSum(arr))
})()

File: String_Array/ThreeSum/three_sum.go
package ThreeSum

import "sort"

func ThreeSum(arr []int) (output [][]int) {
	sort.Ints(arr)

	for i := 0; i < len(arr) - 2; i++ {
		left := i + 1
		right := len(arr) - 1

		if i > 0 && arr[i] == arr[i-1] {
			continue
		}

		for left < right {
			total := arr[i] + arr[left] + arr[right]
			if total == 0 {
				output = append(output, []int{arr[i], arr[left], arr[right]})

				for left < right  && arr[right] == arr[right - 1] {
					right--
				}

				for left < right && arr[left] == arr[left + 1] {
					left++
				}

				left++
				right--
			} else if total > 0 {
				right--
			} else {
				left++
			}
		}
	}

	return
}